test/staking/SwapHelper.test.ts

```diff
@@ -0,0 +1,499 @@
+    await setDAOAddress("ETH_USD_ORACLE", ethUsdOracle.address);
+    await setDAOAddress("GAS_PRICE_ORACLE", gasFeeOracle.address);
+    await setDAOAddress("DAI_ETH_ORACLE", daiEthOracle.address);
+    await setDAOAddress("MARKET_MAKER", marketMaker.address);
+  });
+
+  it("it should swap only safe amount when gains larger than safe amount", async () => {
+    const goodFundManagerFactory = await ethers.getContractFactory(
+      "GoodFundManager"
+    );
+
+    const simpleStaking = await goodCompoundStakingFactory
+      .deploy()
+      .then(async (contract) => {
+        await contract.init(
+          bat.address,
+          cBat.address,
+          nameService.address,
+          "Good BaT",
+          "gBAT",
+          "50",
+          batUsdOracle.address,
+          compUsdOracle.address,
+          [bat.address, dai.address]
+        );
+        return contract;
+      });
+    const reserve = await pair.getReserves();
+    console.log(`reserve ${reserve}`);
+    const currentBlock = await ethers.provider.getBlockNumber();
+    let encodedData = goodFundManagerFactory.interface.encodeFunctionData(
+      "setStakingReward",
+      ["100", simpleStaking.address, currentBlock, currentBlock + 10000, false]
+    );
+    await genericCall(goodFundManager.address, encodedData, avatar, 0);
+    encodedData = goodCompoundStakingFactory.interface.encodeFunctionData(
+      "setcollectInterestGasCostParams",
+      ["250000", "150000"]
+    );
+    await genericCall(simpleStaking.address, encodedData, avatar, 0);
+    await bat["mint(address,uint256)"](
+      founder.address,
+      ethers.utils.parseEther("1001000")
+    );
+    await bat.transfer(cBat.address, ethers.utils.parseEther("1000000")); // We should put extra BAT to mock cBAT contract in order to provide interest
+    await dai["mint(address,uint256)"](
+      founder.address,
+      ethers.utils.parseEther("1000000")
+    );
+    await dai.approve(
+      goodCompoundStaking.address,
+      ethers.utils.parseEther("100")
+    );
+
+    await bat.approve(simpleStaking.address, ethers.utils.parseEther("100"));
+    await simpleStaking.stake(ethers.utils.parseEther("100"), 100, false);
+
+    await cBat.increasePriceWithMultiplier("1500");
+
+    const collectableContracts = await goodFundManager.calcSortedContracts(
+      "1500000"
+    );
+    const currentGains = await simpleStaking.currentGains(true, true);
+    await goodFundManager.collectInterest(collectableContracts, {
+      gasLimit: 1500000,
+    });
+    const currentReserve = await pair.getReserves();
+
+    await simpleStaking.withdrawStake(ethers.utils.parseEther("100"), false);
+    expect(reserve[0].sub(currentReserve[0])).to.be.lt(currentGains[1]);
+
+    encodedData = goodFundManagerFactory.interface.encodeFunctionData(
+      "setStakingReward",
+      ["100", simpleStaking.address, currentBlock, currentBlock + 10000, true]
+    );
+    await genericCall(goodFundManager.address, encodedData, avatar, 0);
```

Comment:
1. move all expects to the end, and make it clear what you are testing
2. it doesnt seem like your expects test what they should.
you want to test that the current gains are X and safe amount is Y, X>Y and collectInterest=Y
so there should be a test that after collection current gains are X-Y, or that the event has value of Y or that uniswap reserve has declined by Y
3. there seem to be lots of unrelevant code in this test
4. do the same for the next test.
5. as usuall any repeating code between tests that IS STILL RELEVANT, you should move to helpers

Reply:
1) Moved expects to the end
2) Improved expects
3)Removed non-relevant codes and moved others ones to helper as much as I could
4)Did
5)Did already as well